lib/service/connect.go

```diff
@@ -862,9 +993,49 @@ func (process *TeleportProcess) rotate(conn *Connector, localState auth.StateV2,
 		return nil, trace.Wrap(err)
 	}
 
+	var assertionID string
+	var systemRoles []types.SystemRole
+	var wantsSystemRoleRepair bool
+	if id.Role == types.RoleInstance {
+		var baseSystemRoles []types.SystemRole
+		for _, baseRole := range conn.ClientIdentity.SystemRoles {
+			baseSystemRoles = append(baseSystemRoles, types.SystemRole(baseRole))
+		}
+		var danglingSystemRoles []types.SystemRole
+		for _, activeRole := range process.getInstanceRoles() {
+			if slices.Contains(baseSystemRoles, activeRole) {
+				continue
+			}
+			danglingSystemRoles = append(danglingSystemRoles, activeRole)
+		}
+
+		if len(danglingSystemRoles) != 0 {
+			process.logger.WarnContext(process.ExitContext(), "rotation logic detected dangling system role(s), will attempt to self-repair", "existing", conn.ClientIdentity.SystemRoles, "dangling", danglingSystemRoles)
+			// if auth is running locally, we can just generate ourselves a new instance cert with the correct
+			// roles. otherwise, we need to prove to the remote auth server that we hold the necessary privileges
+			// by performing system role assertions.
+			if process.getLocalAuth() == nil {
+				assertionID, err = process.assertSystemRoles(danglingSystemRoles)
+				if err != nil && !trace.IsNotImplemented(err) {
+					return nil, trace.Wrap(err)
+				}
+
+				if trace.IsNotImplemented(err) {
+					danglingSystemRoles = nil
+					assertionID = ""
+					process.logger.WarnContext(process.ExitContext(), "auth server does not implement system role assertions, self-repair will be re-attempted later")
```

Comment:
What does "later" mean in this context? I assume it will just keep trying on every startup / `rotate()` interval until the auth server gets updated?

I could be totally off base here but it looks like `RotationConnectionInterval` is 10 seconds by default. Could this end up generating a lot of log spam / network traffic?

Reply:
Good catch! I didn't reevaluate the rotation logic's retry interval when adding this, and it is a lot more aggressive than we like these days.

I've updated the rotation logic to use backoff in its retry, to ensure we don't swamp the auth server (or that if we do swamp it, load abates and we recover quickly).  Rather than try to come up with retry parameters tuned specifically for this usecase I ended up mirroring the retry intervals and configuration used for cache initialization.  Thats been fairly well tested in the wild and I think strikes a good balance between high uptime and minimizing thundering herd issues.